From 4e757eb0ca75ebd3f51bef77a11386f56403fb9e Mon Sep 17 00:00:00 2001
From: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
Date: Fri, 15 Apr 2016 11:54:23 -0400
Bug: https://savannah.nongnu.org/bugs/index.php?47708
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1559933
Subject: [PATCH] Fix missing glyphs in large pcf files.

This partially reverts commit 03ccfd004fcfd3fab4a0c0cea307bdcb0cd79129.

While building grub, a unicode.pf2 font file is generated which is based
off of unifont's /usr/share/fonts/X11/misc/unifont.pcf.gz. That file
contains some 57000+ glyphs, which include glyphs for CJK languages.

Commit 03ccfd00 changed the logic so that only the first 32768 glyphs
would load, which makes grub's font missing some Chinese glyphs, for
example, and show [?] missing glyph boxes instead.

Reverting this minimal chunk from the original commit fixes the issue.

---
 src/pcf/pcfread.c  | 29 +++++++++++++----------------

diff --git a/src/pcf/pcfread.c b/src/pcf/pcfread.c
index afa1480..ee610f4 100644
--- a/src/pcf/pcfread.c
+++ b/src/pcf/pcfread.c
@@ -915,10 +912,10 @@ THE SOFTWARE.
         else
           encodingOffset = FT_GET_SHORT_LE();
 
-        if ( encodingOffset > -1 )
+        if ( encodingOffset != -1 )
         {
           encoding[k].enc   = i * 256 + j;
-          encoding[k].glyph = (FT_UShort)encodingOffset;
+          encoding[k].glyph = (FT_Short)encodingOffset;
 
           FT_TRACE5(( "  code %d (0x%04X): idx %d\n",
                       encoding[k].enc, encoding[k].enc, encoding[k].glyph ));
-- 
2.7.4

